### PR TITLE
DEX-972 Order routes tests

### PR DESCRIPTION
**Pull request by ngadiyak**

### CODE REVIEW

**Review by vsuharnikov — CHANGES_REQUESTED** [review]

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/cancel/CancelAllOrdersSpec.scala [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      r.message.head should have size orders.size
+
+      r.message.foreach(m => {
+        m.foreach(i => {
+          i match {
+            case util.Right(HttpSuccessfulSingleCancel(_,success,status)) => success should be (true); status should be ("OrderCanceled")
+            case _ => fail(s"Unexpected response $r")
+          }
+        })
+      })
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/cancel/CancelAllOrdersSpec.scala:51]: fixed

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/place/package.scala [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import com.wavesplatform.it.MatcherSuiteBase
+import org.scalatest.prop.TableDrivenPropertyChecks
+
+import scala.concurrent.duration.DurationInt
+
+package object place extends MatcherSuiteBase with TableDrivenPropertyChecks {
+
+  val correctExpiration = System.currentTimeMillis() + 10.days.toMillis
+
+  val orderCases = Table(
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/place/package.scala:14]: good idea, fixed

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/place/PlaceLimitOrderSpec.scala [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        dex1.rawApi.place(mkOrder(alice, wavesUsdPair, BUY, 10.waves, 2.usd).json().deepMerge(Json.obj("amount" -> 3.waves))),
+        StatusCodes.BadRequest,
+        9440512,
+        s"The signature of order"
+      )
+    }
+
+    forAll(orderCases) { (n: Int, a: Long, p: Long, f: Long, fa: Asset, t: Long, d: Duration, v: Byte, pk: account.PublicKey, c: Int, e: Int, m: String) =>
+      s"For row $n should return (HTTP-$c; [$e: $m]) " in {
+        validateMatcherErrorContainText(dex1.rawApi.place(mkOrder(alice, wavesUsdPair, BUY, a, p, f, fa, t, d, v, pk)), c, e, m)
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/place/PlaceLimitOrderSpec.scala:49]: ok!

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/place/PlaceMarketOrderSpec.scala [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    "should return error with incorrect order signature" in {
+      validateMatcherErrorContainText(
+        dex1.rawApi.placeMarket(mkOrder(alice, wavesUsdPair, BUY, 10.waves, 2.usd).json().deepMerge(Json.obj("amount" -> 3.waves))),
+        StatusCodes.BadRequest,
+        9440512,
+        s"The signature of order"
+      )
+    }
+
+    forAll(orderCases) { (n: Int, a: Long, p: Long, f: Long, fa: Asset, t: Long, d: Duration, v: Byte, pk: account.PublicKey, c: Int, e: Int, m: String) =>
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/place/PlaceMarketOrderSpec.scala:47]: fixed

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/cancel/CancelOrderSpec.scala [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    "should return error when order already canceled" in {
+      val o = mkOrder(alice, wavesUsdPair, BUY, 10.waves, 1.usd)
+      placeAndAwaitAtDex(o)
+
+      cancelAndAwait(alice, o)
+      validateMatcherError(dex1.rawApi.cancelOrder(alice, o), StatusCodes.BadRequest, 9437194, s"The order ${o.idStr()} is canceled")
+    }
+
+    //TODO: maybe there is should be 404 ?
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/cancel/CancelOrderSpec.scala:50]: ok

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/cancel/CancelOrdersByAddressAndIdsSpec.scala [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      r.message.head should have size orders.size
+
+      r.message.foreach(m => {
+        m.foreach(i => {
+          i match {
+            case util.Right(HttpSuccessfulSingleCancel(_,success,status)) => success should be (true); status should be ("OrderCanceled")
+            case _ => fail(s"Unexpected response $r")
+          }
+        })
+      })
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/cancel/CancelOrdersByAddressAndIdsSpec.scala:64]: ok

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/cancel/CancelOrderByIdSpec.scala [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+      validateMatcherError(
+        dex1.rawApi.cancelOrderById(order.idStr(), Map("X-API-Key" -> apiKey, "X-User-Public-Key" -> "null")),
+        StatusCodes.BadRequest,
+        3148801,
+        "Provided user public key is not correct"
+      )
+    }
+
+    //TODO: maybe there is should be 404?
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/cancel/CancelOrderByIdSpec.scala:81]: ok

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/cancel/CancelOrderByIdSpec.scala [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+    "should return error when order already cancelled" in {
+      val order = mkOrder(alice, wavesUsdPair, BUY, 10.waves, 1.usd)
+      placeAndAwaitAtDex(order)
+
+      dex1.api.cancelOrder(alice, order)
+
+      validateMatcherError(dex1.rawApi.cancelOrderById(order), StatusCodes.BadRequest, 9437194, s"The order ${order.idStr()} is canceled")
+    }
+
+    //TODO: maybe there is should be 404?
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/cancel/CancelOrderByIdSpec.scala:50]: 1024

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/cancel/CancelOrdersByAddressAndIdsSpec.scala [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        mkOrder(alice, wavesUsdPair, BUY, 10.waves, 2.usd),
+        mkOrder(alice, wavesUsdPair, BUY, 10.waves, 3.usd)
+      )
+
+      val ids = orders.map { o =>
+        placeAndAwaitAtDex(o)
+        o.idStr()
+      }
+
+      val r = validate200Json(dex1.rawApi.cancelAllByAddressAndIds(toAddress(alice).stringRepr, ids))
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/cancel/CancelOrdersByAddressAndIdsSpec.scala:51]: done

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/place/PlaceOrderBaseSpec.scala [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+import com.wavesplatform.dex.domain.order.OrderType.BUY
+import com.wavesplatform.dex.it.api.RawHttpChecks
+import com.wavesplatform.dex.it.api.responses.dex.MatcherError
+import com.wavesplatform.dex.it.api.responses.dex.MatcherError.Params
+import com.wavesplatform.it.MatcherSuiteBase
+import org.scalatest.prop.TableDrivenPropertyChecks
+
+import scala.concurrent.duration.DurationInt
+
+class PlaceOrderBaseSpec extends MatcherSuiteBase with TableDrivenPropertyChecks with RawHttpChecks {
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/place/PlaceOrderBaseSpec.scala:15]: 👍

**Review by vsuharnikov — APPROVED** [review]

**Pull request merged**